R/geo.R

```diff
@@ -0,0 +1,217 @@
+#' Geography properties for crunch variables
+#'
+#' Crunch stores geographic data as variable metadata. There are a number of
+#' functions that help access and change this metadata.
+#'
+#' `geo` retrieves the geographic information associate with a variable. If there is geographic information it returns an object of class `CrunchGeography` otherwise it returns `NULL`.
+#'
+#' `CrunchGeography` objects store geography metadata from a variable. There are three slots:
+#' * `geodatum` an object of class CrunchGeodata which stores references to the Crunch-hosted (geo|topo)json to use
+#' * `feature_key` a character that is the feature inside of the (geo|topo)json to match `match_field` to (e.g. properties.name)
+#' * `match_field` a character that is the the variable metadata information to match to `feature_key` to (e.g. name)
+#'
+#' @param x a crunch variable
+#' @param value value of the geography property to set
+#' @return geographic information of class `CrunchGeography` (`NULL` if there is none)
+#'
+#' @name geo
+#'
+#' @examples
+#' \dontrun{
+#' geo(ds$location)
+#'
+#' geo(ds$location)$feature_key <- "properties.name"
+#' geo(ds$location)$match_field <- "name"
+#' }
+#' @aliases geo geo<- CrunchGeography
+NULL
+
+#' @rdname geo
+#' @export
+setMethod("geo", "CrunchVariable", function (x) {
+    var_geodata <- entity(x)@body$view$geodata[[1]]
+    if (is.null(var_geodata)) {
+        # if there's no geodata, return null.
+        return()
+    }
+
+    geodatum <- Geodata(crGET(var_geodata$geodatum))
+    geo_object <- CrunchGeography(
+        geodatum = geodatum,
+        feature_key = var_geodata$feature_key,
+        match_field = var_geodata$match_field
+    )
+
+    return(geo_object)
+})
+#' @rdname geo
+#' @export
+setMethod("geo<-", c("CrunchVariable", "CrunchGeography"),
+          function (x, value) {
+              # if geodatum is of class CrunchGeodata, extact url
```

Comment:
The way I've designed it the class `CrunchGeography` is weakly typed, it should have an object of `GeoData` or a string that is the url to that geodatum in the `geodatum` slot. I don't (yet) have any checks to ensure that's the case, but that's what this is was for. This is helpful for manually specifying geographies (see below) since the url to the geodatum is sufficient for creating the geography view.
```
states_geography <- list(geodatum='https://app.crunch.io/api/geodata/8f9f5fed101042c4815d2dd1fd248cec/',
                         feature_key='properties.name', # parse and list fields?
                         match_field='name') # list possible fields?
states_geography <- new('CrunchGeography', states_geography)

geo(so_data_combined$Country) <- states_geography
```
My plan is to add on an initialization method for `CrunchGeography` that takes the url and makes the appropriate `GeoData` object from it, and then we can remove this totally.